Gift registry queries

sathiyaaa opened the pull request:
## Purpose of this pull request

This pull request (PR) will cover #7994

## Affected DevDocs pages

- /graphql/queries/gift-registry.html
- /graphql/queries/gift-registry-types.html

## Links to Magento source code

-  magento/partners-magento2ee#280

Added queries for giftRegistry and giftRegistryTypes

BarnyShergold commented:
@sathiyaaa - Something doesn't seem right with this PR - you have 3 files changed but 104 commits linked and many belonging to other contributors. Can you check this before the PR is properly reviewed?

shrielenee commented:
@sathiyaaa It looks like you may have not started on the 2.4.2 branch? Please verify you branched off the correct branch.

keharper commented:
@sathiyaaa I've cleaned up this PR so that it is almost ready to merge. I need you to do the following:

1. Fix the problem with too many branches present in this PR.

2. Review all of my edits. There were numerous instances of "Gift card" throughout the PR, and I think I changed them all. I don't have an instance with the GraphQL gift registry functionality, so all of my edits are guesses based on what I can observe from the Admin and on the storefront.

3. Move the content of the `GiftRegistryDynamicAttributeMetadataInterface` section of each topic to a new file in `src/_includes/graphql/gift-registry-dynamic-interface.md`. This file will just contain the intro sentence and the table. Leave the headings in the topic, but replace the moved text with `{% include graphql/gift-registry-dynamic-interface.md %}` See the availableStores query (`src/guides/v2.4/graphql/queries/available-stores.md`) to help visualize what I mean.

4. If the `GiftRegistry` output object will returned in any other query or mutation, move that object and all of its child objects to an include file named `gift-registry.md`.  If that's the case, then the `gift-registry.md` file will contain an include file, and that's OK.

I'm going to need Zilker's help in porting these attribute descriptions into the `schema.graphqls` file. I prefer to leave a round of edits in the code PR, but you created this PR before I could get there.

rogyar commented on src/guides/v2.4/graphql/queries/gift-registry.md:
```diff
@@ -0,0 +1,132 @@
+Attribute | Data type | Description
+--- | --- | ---
+`attribute_group` | String! | Attribute group of the dynamic attribute
+`code` | String! | The internal name of the dynamic attribute
```
`code` attribute has `ID` type in the current implementation
```suggestion
`code` | ID! | The internal name of the dynamic attribute
```

keharper replied:
I didn't cross-check the schema in https://github.com/magento/partners-magento2ee/pull/280. I assumed `code` was the string you see on the Gift Registry Types page in the Admin. If `code` is really an ID, then the description for `code` needs to be updated to reflect that.

rogyar commented on src/guides/v2.4/graphql/queries/gift-registry-types.md:
```diff
@@ -0,0 +1,74 @@
+
+Attribute |  Data Type | Description
+--- | --- | ---
+`dynamic_attributes_metadata` | [GiftRegistryDynamicAttributeMetadataInterface](#giftregistrydynamicattributemetadatainterface) | An array of attributes that define elements of the gift registry. Each attribute is specified as a code-value pair
```
```suggestion
`dynamic_attributes_metadata` | [[GiftRegistryDynamicAttributeMetadataInterface](#giftregistrydynamicattributemetadatainterface)] | An array of attributes that define elements of the gift registry. Each attribute is specified as a code-value pair
```

rogyar reviewed (changes requested):
Hi @sathiyaaa. Great job. Please, check my suggestions below. 
@keharper I see a couple of places where we specify `code` attribute as `string`. But in the current PR it has type `ID` everywhere. I just want to make sure that there were no unpublished schema changes that we are going to use in the future.

sathiyaaa commented:
> @sathiyaaa I've cleaned up this PR so that it is almost ready to merge. I need you to do the following:
> 
> 1. Fix the problem with too many branches present in this PR.
> 2. Review all of my edits. There were numerous instances of "Gift card" throughout the PR, and I think I changed them all. I don't have an instance with the GraphQL gift registry functionality, so all of my edits are guesses based on what I can observe from the Admin and on the storefront.
> 3. Move the content of the `GiftRegistryDynamicAttributeMetadataInterface` section of each topic to a new file in `src/_includes/graphql/gift-registry-dynamic-interface.md`. This file will just contain the intro sentence and the table. Leave the headings in the topic, but replace the moved text with `{% include graphql/gift-registry-dynamic-interface.md %}` See the availableStores query (`src/guides/v2.4/graphql/queries/available-stores.md`) to help visualize what I mean.
> 4. If the `GiftRegistry` output object will returned in any other query or mutation, move that object and all of its child objects to an include file named `gift-registry.md`.  If that's the case, then the `gift-registry.md` file will contain an include file, and that's OK.
> 
> I'm going to need Zilker's help in porting these attribute descriptions into the `schema.graphqls` file. I prefer to leave a round of edits in the code PR, but you created this PR before I could get there.

Sure, will do. Thanks.

keharper commented:
Please include more attributes in the response of both queries. Note that the response for `giftRegistry` does not match the request.

sathiyaaa commented:
Have done the changes. @keharper

keharper commented:
@sathiyaaa I can't merge your PR into `2.4.2-develop` until the code PR is accepted. Instead, I have created a `gift-registry` branch and will merge this PR into that branch. (You'll still get full credit for the contribution). If there are any subsequent schema changes, plus open a PR against the `gift-registry` branch.

sathiyaaa commented:
Sure @keharper, Thank you for reviewing the PR throughout.
